lib/Dialect/FIRRTL/Transforms/GrandCentralSignalMappings.cpp

```diff
@@ -645,18 +631,68 @@ FailureOr<bool> GrandCentralSignalMappingsPass::emitUpdatedMappings(
     op->setAttr("inner_sym", attr);
     return attr;
   };
+
+  auto getSymIdx = [&](Attribute symbol) {
+    auto it = symMap.find(symbol);
+    if (it != symMap.end())
+      return it->second;
+
+    auto id = symbols.size();
+    symbols.push_back(symbol);
+    symMap.insert({symbol, id});
+    return id;
+  };
   auto mkRef = [&](FModuleOp module,
                    const SignalMapping &mapping) -> std::string {
-    if (mapping.localValue.isa<BlockArgument>()) {
-      return llvm::formatv("~{0}|{1}>{2}", circuit.name(), module.getName(),
-                           mapping.localName);
+    SmallString<128> targetRef;
+    auto appendSym = [&](Attribute symbol) {
+      ("{{" + Twine(getSymIdx(symbol)) + "}}").toVector(targetRef);
+    };
+
+    // Start with "circuit", which is top (or DUT?), doesn't matter
+    targetRef = "~";
+    appendSym(FlatSymbolRefAttr::get(pathTop));
+    targetRef.push_back('|');
+
+    // If there's an NLA, emit hierarchical path.
+    if (mapping.nlaSym) {
+      auto nla =
+          cast<HierPathOp>(circuit.lookupSymbol(mapping.nlaSym.getAttr()));
+      assert(!nla.namepath().empty());
+      assert(nla.isComponent());
+
+      // Start from root of NLA, or from top/DUT if through it
+      bool seenRoot = false;
+      bool usesTop = nla.hasModule(pathTop.moduleNameAttr());
+      ArrayRef<Attribute> path = nla.namepath().getValue();
+      for (auto attr : path.drop_back()) {
+        auto ref = attr.cast<hw::InnerRefAttr>();
+        if (usesTop && !seenRoot) {
+          if (ref.getModule() == pathTop.moduleNameAttr())
+            seenRoot = true;
+        }
+
+        if (!usesTop || seenRoot) {
+          appendSym(ref.getModuleRef());
+          targetRef.push_back('/');
+          appendSym(hw::InnerRefAttr::get(ref.getModule(), ref.getName()));
+          targetRef.push_back(':');
+        }
+      };
```

Comment:
This smells like `TokenAnnoTarget.toStr()` or `toVector`.

Reply:
Oh wow, sure thing.  Updated, thanks!